---
title: "Variant_filtering"
author: "Enrico"
date: "16 March 2020"
output: html_document
---

Next step in the pipeline is to filter unwanted variants from the final VCF file.

This are the criteria I will use to filter variants:

(1) Repetitive/Low mappability regions
(2) Indels + Non-biallelic sites
(3) Lynx lynx exclusive substitutions (frequency = 1 in all populations)
(4,5) Hard quality filters, as of GATK standard practices
(6) Under-represented, excessively missing variants
(7) Over/Under covered regions

I will write a script for the first 5 steps which don't require calculations of any filtering thresholds (they are the same no matter the database).
The script is: 2.Variant_filtering-executables/variant_filtering_1to5.sh
```
sbatch variant_filtering_1to5.sh ll_wholegenome_LyCa_ref.sorted
# Submitted batch job 3833293
```
For the last two steps (6 and 7), I will have to write separate script to fine tune my filtering thresholds based on the shape of my data.

To filter out excessively missing variants we have decided the following criteria: we will keep only variants which are represented in ALL populations by at LEAST four individuals. This is because we will be working with allele frequencies and less than four individuals would be really not ideal when calculating allele frequency for a population. Also we would like comparing all populations (are you sure?).

To do this I will write a script that will first divide our VCF of all individuals, into per-population VCFs. I'll first generate a file with the list of all sample names.
```
grep "#CHROM" $LUSTRE/LL_selection/LyCaRef_vcfs/ll_wholegenome_LyCa_ref.sorted.filter5.vcf |
tr '\t' '\n' | grep "_ll_" | cut -d '_' -f1-4 | sort -u \
> $LUSTRE/LL_selection/LyCaRef_bams/all-samples.namelist
```
From this file, I can generate a file with a list of names of each population with a loop. Because Mongolian lynxes are divided into 3 population names (ka,og,to) and I want them as a unique population (mo), I will make an if statement to create a single file for them. The list of rest of the individuals of each population is extracted normally.
```
popARRAY=($(cat $LUSTRE/LL_selection/LyCaRef_bams/all-samples.namelist | cut -d '_' -f3 | sort -u))

for pop in ${popARRAY[@]}
  do
   if [[ ${pop} == ka || ${pop} == og || ${pop} == to ]]
    then
    echo "extracting ${pop} names into mo.namelist"
    grep ${pop} $LUSTRE/LL_selection/LyCaRef_bams/all-samples.namelist \
    >> $LUSTRE/LL_selection/LyCaRef_bams/mo.namelist
   else
    echo "extracting ${pop} names into ${pop}.namelist"
    grep ${pop} $LUSTRE/LL_selection/LyCaRef_bams/all-samples.namelist \
    > $LUSTRE/LL_selection/LyCaRef_bams/${pop}.namelist
   fi
done
```
Then I can use these files to give the names of the samples I want to extract from the VCF file to BCFtools. So I will use BCFtools view in a loop for each population to generate a per-population VCF.

With the per-population VCF, I can again use BCFtools again to filter out variants with missing genotypes for too many samples. By telling BCFtools to include only those variants (-i), the output can be used as a list of positions to exclude from the original VCF.

The script (2.Variant_Filtering-executables/variant_filtering_6.sh) will be launched on CESGA's FT2 server.
```
./variant_filtering_6.sh ll_wholegenome_LyCa_ref.sorted
```
