---
title: "RDA"
author: "Enrico"
date: "15 December 2020"
output: html_document
editor_options:
  chunk_output_type: console
---

In this markdown I will describe how I run Redundancy Analysis (RDA). This analysis uses 2 matrices: (1) per sample genotypes; and (2) per sample bioclimatic variable values (bio5 bio6 bio13 bio14 jan_depth snow_days - ).

Samples that will be excluded from this analysis are:
 - cr_0211 and ba_0233 for lack of coordinates.
 - c_ll_ba_0216, h_ll_ba_0214 and h_ll_ba_0215 for being "bad".

First thing I want to do is to extract the 3 SNPs with the highest BF value from each window, to have a more strict set of SNPs. I will generate a BED of the BayPass results in R which (will be easier) and then upload it to the genomics-b server
```{R}
variables <- c("bio1", "bio2", "bio3", "bio4", "bio5", "bio6", "bio7", "bio8", "bio9", "bio10", "bio11", "bio12", "bio13", "bio14", "bio15", "bio16", "bio17", "bio18", "bio19", "jan_depth", "snow_days")

for (k in 1:length(variables)){

 # Variable:
 var=variables[k]
 # Empty SNPs dataframe to fill
 snps.table <- data.frame()
 
 # Fill SNPs dataframe with all BayPass results for the variable
 for (n in 1:50){
  # upload the dataset table:
  var.snp=read.table(paste0("BayPass_OutPut/AUX_",  var,"_",n,"_summary_betai.out"),h=T)
  
  # calculate the SNP database number sequence (1 every 50) 
  # NOTE: 2100553 is the total number of SNPs in all datasets
  var.snp <- data.frame(var.snp, SNPnum = seq(n, 2100553, by = 50))
  
  # add the dataset rows to the total snps table
  snps.table <- rbind(snps.table, var.snp)
 }
 # order the table based on the SNP number
 snps.table <- snps.table %>% arrange(SNPnum)
 
 # Add SNP ID information from SNPIDs table
 SNPIDs <- read_tsv("BayPass_OutPut/finalset.maf5pc.SNPIDs", col_names = F)[,2-3] %>%
    rename("scaffold" =  X2, "position" = X3)
  
 # Add SNP IDs to the total snps table
 snps.table <- data.frame(snps.table,SNPIDs)

 # Results BED file:
 bedresults <- data.frame(chr=snps.table$scaffold, start=(snps.table$position -1),
                          stop=snps.table$position, BF=snps.table$BF.dB.)
 write.table(x = bedresults,
             file = paste0("BayPass_results/",var,"_snps_BFs.bed"),
             quote=FALSE,  col.names = F, row.names = FALSE, sep= "\t")
}
```
Now to extract the top 3 SNPs for each outlier window and take only top 3 from "inversion" region
located between 17195000 and 18625000 of scaffold_17_arrow_ctg1. To only have one window (row in the table) in this region, I manually modified the "GenWin_windows_outliers.tsv" files. The new version with only one row for the inversion is located in the RDA folder (GenWin_windows_outliers_onesuper.tsv).

```{R}
for (k in 1:length(variables)){

 # Variable:
 var=variables[k]

 bedresults <- read.table(paste0("BayPass_results/",var,"_snps_BFs.bed"),h=F)
 outlier_windows <- read.table(paste0("RDA/",var,"_GenWin_windows_outliers_onesuper.tsv"),
                               h=T, as.is = T)
 topsnps <- data.frame()
 for (i in 1:nrow(outlier_windows)){
   window <- outlier_windows[i,]
   scaffold <- window$scaffold
   start <- window$WindowStart
   stop <- window$WindowStop
   window_snps <- subset(bedresults, V1 == scaffold & V3 >= start & V3 <= stop)
   topsnp <- window_snps[order(window_snps$V4,decreasing = TRUE),]
   topsnp <- topsnp[1:3,]
   topsnp$var <- var
   topsnp$V4 <- NULL
   topsnps <- rbind(topsnps, topsnp)
 }
 write.table(x = topsnps,
             file = paste0("RDA/",var,"_topsnps.range"),
             quote=FALSE,  col.names = F, row.names = FALSE, sep= "\t")
}
```
Copy the top snps table (plink range format) to genomics-b server
```{bash}
scp Documents/Selection_Eurasian_Lynx/RDA/*_topsnps.range ebazzicalupo@genomics-b.ebd.csic.es:/home/ebazzicalupo/RDA
```
The package I'm going to use for RDA (vegan) can use an input format for the genotype matrix (raw) which is generated by PLINK. To generate this in plink:
```{bash}
cd /home/ebazzicalupo/RDA

# First generate the range file for the top SNPs of the chosen variables (see Corr_PCA_variables.R):
# bio1 bio2 bio3 bio4 bio5 bio6 bio7 bio8 bio9 bio10 bio11 bio12 bio13 bio14 bio15 bio16 bio17 bio18 bio19 jan_depth snow_days
for var in bio1 bio5 bio7 bio12 bio13 jan_depth snow_days
 do
  echo ${var}
  cat ${var}_topsnps.range >> tmp
done
sort -k1,1 -k2,2n tmp | uniq > sixvars_topsnps.range && rm tmp

# I manually created a file listing samples to remove for plink with following format:
# c_ll_ba_0216 c_ll_ba_0216 0 0 0 -9
# c_ll_ba_0233 c_ll_ba_0233 0 0 0 -9
# c_ll_cr_0211 c_ll_cr_0211 0 0 0 -9
# h_ll_ba_0214 h_ll_ba_0214 0 0 0 -9
# h_ll_ba_0215 h_ll_ba_0215 0 0 0 -9

VCF=/home/ebazzicalupo/BayPass/VCF/ll_wholegenome_LyCa_ref.sorted.filter7.vcf

plink_1.9 --vcf ${VCF} --double-id --allow-extra-chr --set-missing-var-ids @:# \
--extract range sixvars_topsnps.range \
--remove samplestoremove.txt --geno 0.001 \
--recode A --out sixvar_topsnps_persample
```
Copy the .raw file to laptop:
```{bash}
scp ebazzicalupo@genomics-b.ebd.csic.es:~/RDA/sixvar_topsnps_persample.raw ~/Documents/Selection_Eurasian_Lynx/RDA/
```
Prepare R:
```{R}
library(tidyverse)
library(viridis)
library(RColorBrewer)
library(vegan)
library(adegenet)
```
In R
```{R}
# Load GenoType Data
gt_data <- read.PLINK("RDA/sixvar_topsnps_persample.raw")
gt_data_tsv <- read_delim("RDA/sixvar_topsnps_persample.raw", delim = ' ')

# Load environmental data (select columns -> N = bioN + 1)
sixvar_env <- read_tsv("WorldClim_table_persample.tsv", col_names = T)[,c(1,2,4,6,8,9,13,14,16)] %>%
  column_to_rownames(., var="sample")
# Add snow data
sixvar_env <- cbind(sixvar_env, read_tsv("Snow_table_persample.tsv", col_names = T)[,-1])

# Run RDA
rda <- rda(gt_data ~ ., data=sixvar_env, scale=T)

RsquareAdj(rda)
summary(eigenvals(rda, model = "constrained"))
signif.full <- anova.cca(rda, parallel=getOption("mc.cores")) # default is permutation=999
signif.full

loc <- rep(NA, length(gt_data_tsv$FID))
loc[grep("ba", gt_data_tsv$FID)] <- "Balkans"
loc[grep("ca", gt_data_tsv$FID)] <- "Caucasus"
loc[grep("cr", gt_data_tsv$FID)] <- "Carpathians"
loc[grep("ka", gt_data_tsv$FID)] <- "Mongolia"
loc[grep("ki", gt_data_tsv$FID)] <- "Kirov"
loc[grep("la", gt_data_tsv$FID)] <- "Latvia"
loc[grep("no", gt_data_tsv$FID)] <- "Norway"
loc[grep("po", gt_data_tsv$FID)] <- "NE-Poland"
loc[grep("og", gt_data_tsv$FID)] <- "Mongolia"
loc[grep("to", gt_data_tsv$FID)] <- "Mongolia"
loc[grep("tu", gt_data_tsv$FID)] <- "Tuva"
loc[grep("ur", gt_data_tsv$FID)] <- "Urals"
loc[grep("vl", gt_data_tsv$FID)] <- "Vladivostok"
loc[grep("ya", gt_data_tsv$FID)] <- "Yakutia"

num <- rep(NA, length(gt_data_tsv$FID))
num[grep("ba", gt_data_tsv$FID)] <- 1
num[grep("ca", gt_data_tsv$FID)] <- 3
num[grep("cr", gt_data_tsv$FID)] <- 2
num[grep("ka", gt_data_tsv$FID)] <- 6
num[grep("ki", gt_data_tsv$FID)] <- 4
num[grep("la", gt_data_tsv$FID)] <- 5
num[grep("no", gt_data_tsv$FID)] <- 8
num[grep("po", gt_data_tsv$FID)] <- 7
num[grep("og", gt_data_tsv$FID)] <- 6
num[grep("to", gt_data_tsv$FID)] <- 6
num[grep("tu", gt_data_tsv$FID)] <- 9
num[grep("ur", gt_data_tsv$FID)] <- 10
num[grep("vl", gt_data_tsv$FID)] <- 11
num[grep("ya", gt_data_tsv$FID)] <- 12


ola <- data.frame(sample = gt_data_tsv$FID, pop = loc, n = num)
eco <- levels(ola$pop)
bg <- cols <- c("#A035AF",
                brewer.pal(12,"Paired")[9],
                "#B8860b",
                viridis_pal()(5)[1],
                brewer.pal(12,"Paired")[3],
                brewer.pal(12,"Paired")[7],
                viridis_pal()(5)[3],
                viridis_pal()(5)[2],
                brewer.pal(12,"Paired")[8],
                "#0F4909",
                brewer.pal(12,"Paired")[5],
                brewer.pal(12,"Paired")[6])


plot(rda, type="n", scaling=3)
#points(rda, display="species", pch=20, cex=0.7, col="gray32", scaling=3)           # the SNPs
points(rda, display="sites", pch=21, cex=1.3, col="gray32", scaling=3, bg=bg[num]) # the wolves
text(rda, scaling=3, display="bp", col="#0868ac", cex=1)     # the predictors
legend("bottomright", legend=eco, bty="n", col="gray32", pch=21, cex=0.7, pt.bg=bg)

```

To run RDA on neutral regions of the genome:
```{bash}
# Generate neutral range file:
bed=/home/ebazzicalupo/co-inertia/12k10k_neutral.bed
name=12k10k_neutral
nlines=($(wc -l < ${bed}))
paste ${bed} <(yes ${name} | head -${nlines}) > /home/ebazzicalupo/RDA/${name}.range

# generate raw file of neutral regions
cd /home/ebazzicalupo/RDA

VCF=/home/ebazzicalupo/BayPass/VCF/ll_wholegenome_LyCa_ref.sorted.filter7.vcf

plink_1.9 --vcf ${VCF} --double-id --allow-extra-chr --set-missing-var-ids @:# \
--extract range ${name}.range \
--remove samplestoremove.txt --geno 0.001 \
--recode A --out ${name}_persample
```
Copy to laptop neutral raw file:
```{bash}
scp ebazzicalupo@genomics-b.ebd.csic.es:~/RDA/12k10k_neutral_persample.raw ~/Documents/Selection_Eurasian_Lynx/RDA/
```
Prepare R:
```{R}
library(tidyverse)
library(viridis)
library(RColorBrewer)
library(vegan)
library(adegenet)
```
In R
```{R}
# Load GenoType Data
gt_data <- read.PLINK("RDA/12k10k_neutral_persample.raw")
gt_data_tsv <- read_delim("RDA/12k10k_neutral_persample.raw", delim = ' ')

# Load environmental data (select columns -> N = bioN + 1)
sixvar_env <- read_tsv("WorldClim_table_persample.tsv", col_names = T)[,c(1,2,6,8,13,14)] %>%
  column_to_rownames(., var="sample")
# Add snow data
sixvar_env <- cbind(sixvar_env, read_tsv("Snow_table_persample.tsv", col_names = T)[,-1])

# Run RDA
rda <- rda(gt_data ~ ., data=sixvar_env, scale=T)

RsquareAdj(rda)
summary(eigenvals(rda, model = "constrained"))
signif.full <- anova.cca(rda, parallel=getOption("mc.cores")) # default is permutation=999
signif.full

loc <- rep(NA, length(gt_data_tsv$FID))
loc[grep("ba", gt_data_tsv$FID)] <- "Balkans"
loc[grep("ca", gt_data_tsv$FID)] <- "Caucasus"
loc[grep("cr", gt_data_tsv$FID)] <- "Carpathians"
loc[grep("ka", gt_data_tsv$FID)] <- "Mongolia"
loc[grep("ki", gt_data_tsv$FID)] <- "Kirov"
loc[grep("la", gt_data_tsv$FID)] <- "Latvia"
loc[grep("no", gt_data_tsv$FID)] <- "Norway"
loc[grep("po", gt_data_tsv$FID)] <- "NE-Poland"
loc[grep("og", gt_data_tsv$FID)] <- "Mongolia"
loc[grep("to", gt_data_tsv$FID)] <- "Mongolia"
loc[grep("tu", gt_data_tsv$FID)] <- "Tuva"
loc[grep("ur", gt_data_tsv$FID)] <- "Urals"
loc[grep("vl", gt_data_tsv$FID)] <- "Vladivostok"
loc[grep("ya", gt_data_tsv$FID)] <- "Yakutia"

num <- rep(NA, length(gt_data_tsv$FID))
num[grep("ba", gt_data_tsv$FID)] <- 1
num[grep("ca", gt_data_tsv$FID)] <- 3
num[grep("cr", gt_data_tsv$FID)] <- 2
num[grep("ka", gt_data_tsv$FID)] <- 6
num[grep("ki", gt_data_tsv$FID)] <- 4
num[grep("la", gt_data_tsv$FID)] <- 5
num[grep("no", gt_data_tsv$FID)] <- 8
num[grep("po", gt_data_tsv$FID)] <- 7
num[grep("og", gt_data_tsv$FID)] <- 6
num[grep("to", gt_data_tsv$FID)] <- 6
num[grep("tu", gt_data_tsv$FID)] <- 9
num[grep("ur", gt_data_tsv$FID)] <- 10
num[grep("vl", gt_data_tsv$FID)] <- 11
num[grep("ya", gt_data_tsv$FID)] <- 12


ola <- data.frame(sample = gt_data_tsv$FID, pop = loc, n = num)
eco <- levels(ola$pop)
bg <- cols <- c("#A035AF",
                brewer.pal(12,"Paired")[9],
                "#B8860b",
                viridis_pal()(5)[1],
                brewer.pal(12,"Paired")[3],
                brewer.pal(12,"Paired")[7],
                viridis_pal()(5)[3],
                viridis_pal()(5)[2],
                brewer.pal(12,"Paired")[8],
                "#0F4909",
                brewer.pal(12,"Paired")[5],
                brewer.pal(12,"Paired")[6])

pdf(file = "RDA_12k10k_neutral.pdf")
plot(rda, type="n", scaling=3)
#points(rda, display="species", pch=20, cex=0.7, col="gray32", scaling=3)           # the SNPs
points(rda, display="sites", pch=21, cex=1.3, col="gray32", scaling=3, bg=bg[num]) # the wolves
text(rda, scaling=3, display="bp", col="#0868ac", cex=1)     # the predictors
legend("bottomright", legend=eco, bty="n", col="gray32", pch=21, cex=0.7, pt.bg=bg)

dev.off()

```

To run RDA on combined differentiation candidate regions of selection:
```{bash}
```
